exchanges/exchanges.go

```diff
@@ -1821,9 +1956,367 @@ func (r PoloniexCandlestickResponse) translate(bin candlestickKey) Candlesticks
+}
+
+// Poloniex's WebsocketProcessor. Handles messages of type "i", "o", and "t".
+func (poloniex *PoloniexExchange) processWsMessage(raw []byte) {
+	msg := make([]interface{}, 0)
+	err := json.Unmarshal(raw, &msg)
+	if err != nil {
+		poloniex.setWsFail(err)
+		return
+	}
+	if len(msg) == 1 {
+		// Likely a heatbeat
+		code, ok := msg[0].(float64)
+		if !ok {
+			poloniex.setWsFail(fmt.Errorf("non-integer single-element poloniex response of implicit type %T", msg[0]))
+			return
+		}
+		intCode := int(code)
+		if intCode == poloniexHeartbeatCode {
+			return
+		}
+		poloniex.setWsFail(fmt.Errorf("unknown code in single-element poloniex response: %d", intCode))
+		return
+	}
+	if len(msg) < 3 {
+		poloniex.setWsFail(fmt.Errorf("poloniex websocket message had unexpected length %d", len(msg)))
+		return
+	}
+
+	responseList, ok := msg[2].([]interface{})
+	if !ok {
+		poloniex.setWsFail(fmt.Errorf("poloniex websocket message type assertion failure: %T", msg[2]))
+		return
+	}
+
+	code := firstCode(responseList)
+
+	if code == poloniexInitialOrderbookKey {
+		poloniex.processWsOrderbook(responseList)
+		state := poloniex.state()
+		if state != nil { // Only send update if price has been fetched
+			depth := poloniex.wsDepths()
+			poloniex.Update(&ExchangeState{
+				Price:        state.Price,
+				BaseVolume:   state.BaseVolume,
+				Volume:       state.Volume,
+				Change:       state.Change,
+				Depth:        depth,
+				Candlesticks: state.Candlesticks,
+			})
+		}
+		return
+	}
+
+	if code != poloniexOrderUpdateKey && code != poloniexTradeUpdateKey {
+		poloniex.setWsFail(fmt.Errorf("Unexpected code in first element of poloniex websocket response list: %s", code))
+		return
+	}
+
+	poloniex.orderMtx.RLock()
+	newAsks := make(poloniexOrders)
+	newBids := make(poloniexOrders)
+	for _, update := range responseList {
+		updateParams, ok := update.([]interface{})
+		if !ok {
+			poloniex.setWsFail(fmt.Errorf("failed to type convert poloniex orderbook update array"))
+			return
+		}
+		if len(updateParams) < 4 {
+			poloniex.setWsFail(fmt.Errorf("unexpected number of parameters in poloniex orderboook update"))
+			return
+		}
+		updateType, ok := updateParams[0].(string)
+		if !ok {
+			poloniex.setWsFail(fmt.Errorf("failed to type convert poloniex orderbook update type"))
+			return
+		}
+
+		var order *poloniexOrder
+		var direction int
+		if updateType == poloniexOrderUpdateKey {
+			order, direction, err = processPoloniexOrderbookUpdate(updateParams)
+			if err != nil {
+				poloniex.setWsFail(err)
+			}
+		} else if updateType == poloniexTradeUpdateKey {
+			continue
+			// trade, direction, err = poloniex.processTrade(updateParams)
+		}
+
+		if direction == poloniexAskDirection {
+			newAsks[int64(order.price*1e8)] = order
+		} else if direction == poloniexBuyDirection {
+			newBids[int64(order.price*1e8)] = order
+		} else {
+			poloniex.setWsFail(fmt.Errorf("Unknown poloniex update direction indicator: %d", direction))
+			return
```

Comment:
Would def go for a switch on `direction` for these.  It's generally considered good practice in go to prefer a switch over if-else if-else chains when appropriate.